<?xml version="1.0"?>
<sdf version="1.6">
  <model name="usv">
    <link name="base_link">
      <inertial>
        <pose>0 0 -0.2 0 0 0</pose>
        <mass>10</mass>
        <inertia>
          <ixx>0.5666666666666668</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>1.0416666666666667</iyy>
          <iyz>0</iyz>
          <izz>1.5416666666666667</izz>
        </inertia>
      </inertial>
      <visual name="visual">
        <geometry>
           <mesh>
            <uri>file://usv/meshes/usv.dae</uri>
            <scale>0.1 0.1 0.1</scale>
          </mesh>
        </geometry>
        <!-- <material>
          <diffuse>1.0 1.0 1.0</diffuse>
          <specular>1.0 1.0 1.0</specular> -->
          <!-- <pbr>
            <metal>
              <albedo_map>file://usv/materials/textures/Boat01_Albedo.png</albedo_map>
              <normal_map>file://usv/materials/textures/Boat01_Normal.png</normal_map>
              <roughness_map>file://usv/materials/textures/Boat01_Roughness.png</roughness_map>
              <metalness_map>file://usv/materials/textures/Boat01_Metallic.png</metalness_map>
            </metal>
          </pbr> -->
          <!-- fallback to script if no PBR support-->
          <!-- <script>
            <uri>file://usv/materials/scripts/</uri>
            <uri>file://usv/materials/textures/</uri>
            <name>Fallback/Boat01_Diffuse</name>
          </script>
        </material> -->
      </visual>

      <collision name="Box011">
        <pose>-0.020046 0.204338 -0.1352701 0.0 0.0 0.0</pose>
        <geometry>
          <box>
            <size>0.89985 0.273203 0.232172</size>
          </box>
        </geometry>
      </collision>
      <collision name="Box012">
        <pose>-0.020046 -0.20434 -0.1352701 0.0 0.0 0.0</pose>
        <geometry>
          <box>
            <size>0.89985 0.273203 0.232172</size>
          </box>
        </geometry>
      </collision>
      <collision name="Box013">
        <pose>0.0099115 0.0 0.060029 0.0 0.0 0.0</pose>
        <geometry>
          <box>
            <size>0.998409 0.789244 0.4260934</size>
          </box>
        </geometry>
      </collision>

      <sensor name="navsat" type="navsat">
        <pose>0.3 0.0 0.25 0 0 0</pose>
        <always_on>1</always_on>
        <update_rate>1</update_rate>
      </sensor>

      <sensor name="imu" type="imu">
        <pose>0.2 0.0 0.25 0 0 0</pose>
        <always_on>1</always_on>
        <update_rate>1</update_rate>
      </sensor>

      <sensor name='gpu_lidar' type='gpu_lidar'>
        <pose>0.2 0.0 0.55 0 0 0</pose>
        <topic>lidar</topic>
        <update_rate>10</update_rate>
        <lidar>
          <scan>
            <horizontal>
              <samples>640</samples>
              <resolution>1</resolution>
              <min_angle>-3.14</min_angle>
              <max_angle>3.14</max_angle>
            </horizontal>
            <vertical>
              <samples>16</samples>
              <resolution>1</resolution>
              <min_angle>-0.261799</min_angle>
              <max_angle>0.261799</max_angle>
            </vertical>
          </scan>
          <range>
            <min>0.08</min>
            <max>10.0</max>
            <resolution>0.01</resolution>
          </range>
        </lidar>
        <alwaysOn>1</alwaysOn>
        <visualize>true</visualize>
        <plugin
          filename="RosGzPointCloud"
          name="ros_gz_point_cloud::PointCloud">
          <namespace>custom_params</namespace>
          <topic>pc2</topic>
          <frame_id>custom_params/link</frame_id>
        </plugin>
      </sensor>

      <sensor name="rgbd_camera" type="rgbd_camera">
        <pose>0.55 0 0.25 0 0 0</pose>
        <camera>
          <horizontal_fov>1.047</horizontal_fov>
          <image>
            <width>320</width>
            <height>240</height>
          </image>
          <clip>
            <near>0.1</near>
            <far>100</far>
          </clip>
        </camera>
        <always_on>1</always_on>
        <update_rate>30</update_rate>
        <visualize>true</visualize>
        <topic>zed_rgbd</topic>
        <plugin
          filename="RosGzPointCloud"
          name="ros_gz_point_cloud::PointCloud">
          <namespace>custom_params</namespace>
          <topic>pc2</topic>
          <frame_id>map</frame_id>
        </plugin>
      </sensor>

    </link>

    <!-- Left thruster housing -->
    <link name="left_housing_link">
      <inertial>
        <pose>-0.3 0.23 -0.31 0 0 0</pose>
        <mass>1</mass>
        <inertia>
          <ixx>0.0020290833333333333</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.002408333333333333</iyy>
          <iyz>0</iyz>
          <izz>0.0006874166666666666</izz>
        </inertia>
      </inertial>

      <visual name="left_housing_visual">
        <pose>-0.3 0.23 -0.31 0 0 0</pose>
        <geometry>
           <mesh>
            <uri>file://usv/meshes/housing.dae</uri>
          </mesh>
        </geometry>
        <material>
          <diffuse>1.0 1.0 1.0</diffuse>
          <specular>1.0 1.0 1.0</specular>
          <pbr>
            <metal>
              <albedo_map>file://usv/materials/textures/Thruster_Albedo.png</albedo_map>
              <roughness_map>file://usv/materials/textures/Thruster_Roughness.png</roughness_map>
            </metal>
          </pbr>
        </material>
      </visual>
  
      <collision name="Cylinder263d">
        <pose>-0.2833463 0.22986701 -0.31 1.5708 0.0 -1.5708</pose>
        <geometry>
          <cylinder>
            <length>0.14376</length>
            <radius>0.0243692</radius>
          </cylinder>
        </geometry>
      </collision>
    </link>

    <!-- Left propeller -->
    <link name="left_prop_link">
      <inertial>
        <pose>-0.300872875 0.229867016 -0.31 1.5708 0.0 -1.5708</pose>
        <mass>0.1</mass>
        <inertia>
          <ixx>0.00010562499999999999</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.00010562499999999999</iyy>
          <iyz>0</iyz>
          <izz>0.00015125000000000002</izz>
        </inertia>
      </inertial>
      <visual name="left_prop_visual">
        <pose>-0.3 0.23 -0.31 0 0 0</pose>
        <geometry>
           <mesh>
            <uri>file://usv/meshes/prop.dae</uri>
          </mesh>
        </geometry>
        <material>
          <diffuse>1.0 1.0 1.0</diffuse>
          <specular>1.0 1.0 1.0</specular>
          <pbr>
            <metal>
              <albedo_map>file://usv/materials/textures/Thruster_Albedo.png</albedo_map>
              <roughness_map>file://usv/materials/textures/Thruster_Roughness.png</roughness_map>
            </metal>
          </pbr>
        </material>
      </visual>

      <collision name="Cylinder262d">
        <pose>-0.300872875 0.229867016 -0.31 1.5708 0.0 -1.5708</pose>
        <geometry>
          <cylinder>
            <length>0.0608172</length>
            <radius>0.0553844</radius>
          </cylinder>
        </geometry>
      </collision>
    </link>

    <!-- Right thruster housing -->
    <link name="right_housing_link">
      <inertial>
        <pose>-0.3 -0.23 -0.31 0 0 0</pose>
        <mass>1</mass>
        <inertia>
          <ixx>0.0020290833333333333</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.002408333333333333</iyy>
          <iyz>0</iyz>
          <izz>0.0006874166666666666</izz>
        </inertia>
      </inertial>
      <visual name="right_housing_visual">
        <pose>-0.3 -0.23 -0.31 0 0 0</pose>
        <geometry>
           <mesh>
            <uri>file://usv/meshes/housing.dae</uri>
          </mesh>
        </geometry>
        <material>
          <diffuse>1.0 1.0 1.0</diffuse>
          <specular>1.0 1.0 1.0</specular>
          <pbr>
            <metal>
              <albedo_map>file://usv/materials/textures/Thruster_Albedo.png</albedo_map>
              <roughness_map>file://usv/materials/textures/Thruster_Roughness.png</roughness_map>
            </metal>
          </pbr>
        </material>
      </visual>

      <!-- Thruster collisions -->
      <collision name="Cylinder263i">
        <pose>-0.2833463 -0.22986701 -0.31 1.5708 0.0 -1.5708</pose>
        <geometry>
          <cylinder>
            <length>0.14376</length>
            <radius>0.0243692</radius>
          </cylinder>
        </geometry>
      </collision>
    </link>

    <!-- Right propeller -->
    <link name="right_prop_link">
      <inertial>
        <pose>-0.300872875 -0.229867016 -0.31 1.5708 0.0 -1.5708</pose>
        <mass>0.1</mass>
        <inertia>
          <ixx>0.00010562499999999999</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>0.00010562499999999999</iyy>
          <iyz>0</iyz>
          <izz>0.00015125000000000002</izz>
        </inertia>
      </inertial>
      <visual name="right_prop_visual">
        <pose>-0.3 -0.23 -0.31 0 0 0</pose>
        <geometry>
           <mesh>
            <uri>file://usv/meshes/prop.dae</uri>
          </mesh>
        </geometry>
        <material>
          <diffuse>1.0 1.0 1.0</diffuse>
          <specular>1.0 1.0 1.0</specular>
          <pbr>
            <metal>
              <albedo_map>file://usv/materials/textures/Thruster_Albedo.png</albedo_map>
              <roughness_map>file://usv/materials/textures/Thruster_Roughness.png</roughness_map>
            </metal>
          </pbr>
        </material>
      </visual>

      <collision name="Cylinder262i">
        <pose>-0.300872875 -0.229867016 -0.31 1.5708 0.0 -1.5708</pose>
        <geometry>
          <cylinder>
            <length>0.0608172</length>
            <radius>0.0553844</radius>
          </cylinder>
        </geometry>
      </collision>
    </link>

    <!-- Chassis-left engine joint -->
    <joint name="left_chasis_engine_joint" type="revolute">
      <axis>
        <xyz>0 0 1</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
          <effort>10</effort>
          <velocity>10</velocity>
        </limit>
      </axis>
      <parent>base_link</parent>
      <child>left_housing_link</child>
    </joint>

    <!-- Left joint housing-propeller -->
    <joint name="left_engine_propeller_joint" type="revolute">
      <pose>-0.300872875 0.229867016 -0.31 1.5708 0.0 0</pose>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <effort>100</effort>
          <velocity>100</velocity>
        </limit>
        <dynamics>
          <friction>0.05</friction>
          <damping>0.05</damping>
        </dynamics>
      </axis>
      <parent>left_housing_link</parent>
      <child>left_prop_link</child>
    </joint>

    <!-- Chassis-right engine joint -->
    <joint name="right_chasis_engine_joint" type="revolute">
      <axis>
        <xyz>0 0 1</xyz>
        <limit>
          <lower>0</lower>
          <upper>0</upper>
          <effort>10</effort>
          <velocity>10</velocity>
        </limit>
      </axis>
      <parent>base_link</parent>
      <child>right_housing_link</child>
    </joint>

    <!-- Right joint housing-propeller -->
    <joint name="right_engine_propeller_joint" type="revolute">
      <pose>-0.300872875 -0.229867016 -0.31 1.5708 0.0 0</pose>
      <axis>
        <xyz>1 0 0</xyz>
        <limit>
          <effort>100</effort>
          <velocity>100</velocity>
        </limit>
        <dynamics>
          <friction>0.05</friction>
          <damping>0.05</damping>
        </dynamics>
      </axis>
      <parent>right_housing_link</parent>
      <child>right_prop_link</child>
    </joint>

    <!-- Left hull -->
    <plugin
      filename="libSurface.so"
      name="vrx::Surface">
      <link_name>base_link</link_name>
      <hull_length>1.15</hull_length>
      <hull_radius>0.12</hull_radius>
      <fluid_level>0.1</fluid_level>
      <points>
        <point>0.68 0.204338 -0.1552701</point>
        <point>-0.465 0.204338 -0.1552701</point>
      </points>
      <wavefield>
        <topic>/vrx/wavefield/parameters</topic>
      </wavefield>
    </plugin>

    <!-- Right hull -->
    <plugin
      filename="libSurface.so"
      name="vrx::Surface">
      <link_name>base_link</link_name>
      <hull_length>1.15</hull_length>
      <hull_radius>0.12</hull_radius>
      <fluid_level>0.1</fluid_level>
      <points>
        <point>0.68 -0.204338 -0.1552701 </point>
        <point>-0.465 -0.204338 -0.1552701 </point>
      </points>
      <wavefield>
        <topic>/vrx/wavefield/parameters</topic>
      </wavefield>
    </plugin>

    <plugin
      filename="libSimpleHydrodynamics.so"
      name="vrx::SimpleHydrodynamics">
      <link_name>base_link</link_name>
      <!-- Added mass -->
      <xDotU>0.0</xDotU>
      <yDotV>0.0</yDotV>
      <nDotR>0.0</nDotR>
      <!-- Linear and quadratic drag -->
      <xU>4.0</xU>
      <xUU>6.0</xUU>
      <yV>4.0</yV>
      <yVV>4.0</yVV>
      <zW>20.0</zW>
      <kP>12.0</kP>
      <kPP>24.0</kPP>
      <mQ>36.0</mQ>
      <mQQ>35.0</mQQ>
      <nR>32.0</nR>
      <nRR>32.0</nRR>
    </plugin>

    <plugin
      filename="gz-sim-thruster-system"
      name="gz::sim::systems::Thruster">
      <joint_name>left_engine_propeller_joint</joint_name>
      <thrust_coefficient>0.0001</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
      <velocity_control>true</velocity_control>
    </plugin>

    <plugin
      filename="gz-sim-thruster-system"
      name="gz::sim::systems::Thruster">
      <joint_name>right_engine_propeller_joint</joint_name>
      <thrust_coefficient>0.0001</thrust_coefficient>
      <fluid_density>1000</fluid_density>
      <propeller_diameter>0.1</propeller_diameter>
      <velocity_control>true</velocity_control>
    </plugin>
  </model>
</sdf>

<!--

ros2 launch vrx_gz vrx_environment.launch.py world:=nbpark

ros2 run ros_gz_bridge parameter_bridge /model/my_roboboat/joint/left_engine_propeller_joint/cmd_thrust@std_msgs/msg/Float64]gz.msgs.Double
ros2 run ros_gz_bridge parameter_bridge /model/my_roboboat/joint/right_engine_propeller_joint/cmd_thrust@std_msgs/msg/Float64]gz.msgs.Double

ros2 topic pub /model/my_roboboat/joint/left_engine_propeller_joint/cmd_thrust std_msgs/msg/Float64 "data: 150"
ros2 topic pub /model/my_roboboat/joint/right_engine_propeller_joint/cmd_thrust std_msgs/msg/Float64 "data: 150"

-->